agent/agents/mongodb/mongolog/internal/mongolog.go

```diff
@@ -0,0 +1,273 @@
+	labels := pprof.Labels("component", "mongodb.mongolog")
+
+	// create new session
+	client, err := createSession(l.mongoDSN, l.agentID)
+	if err != nil {
+		return err
+	}
+	logsPath, err := getLogFilePath(client)
+	if err != nil {
+		return err
+	}
+	err = client.Disconnect(ctx)
+	if err != nil {
+		l.logger.Warningln(err)
+	}
+
+	// create aggregator which collects documents and aggregates them into qan report
+	l.aggregator = aggregator.New(time.Now(), l.agentID, l.logger, l.maxQueryLength)
+	reportChan := l.aggregator.Start()
+
+	// create sender which sends qan reports and start it
+	l.sender = sender.New(reportChan, l.w, l.logger)
+	err = l.sender.Start()
+	if err != nil {
+		return err
+	}
+
+	// create new channel over which
+	// we will tell goroutine it should close
+	l.doneChan = make(chan struct{})
+
+	// start a goroutine and Add() it to WaitGroup
+	// so we could later Wait() for it to finish
+	l.wg = &sync.WaitGroup{}
+	l.wg.Add(2)
+
+	// create ready sync.Cond so we could know when goroutine actually started getting data from db
+	ready := sync.NewCond(&sync.Mutex{})
+	ready.L.Lock()
+	defer ready.L.Unlock()
+
+	logsPathWithPrefix := path.Join(l.logFilePrefix, logsPath)
+	reader, err := filereader.NewContinuousFileReader(logsPathWithPrefix, l.logger)
+	if err != nil {
+		return err
+	}
+	// create monitors service which we use to periodically scan server for new/removed databases
+	l.monitor = NewMonitor(logsPathWithPrefix, reader, l.logger)
+
+	go pprof.Do(ctx, labels, func(ctx context.Context) {
+		start(ctx, l.monitor, l.aggregator, l.wg, l.doneChan, ready, l.logger)
+	})
+
+	// wait until we actually fetch data from db
+	ready.Wait()
+
+	l.running = true
+	return nil
+}
+
+// Stop stops running mongolog, waits until it stops.
+func (l *Mongolog) Stop() error {
+	l.m.Lock()
+	defer l.m.Unlock()
+	if !l.running {
+		return nil
+	}
+
+	// notify goroutine to close
+	close(l.doneChan)
+	l.monitor.Stop()
+
+	// wait for goroutine to exit
+	l.wg.Wait()
+
+	// set state to "not running"
+	l.running = false
+	return nil
+}
+
+func start(ctx context.Context, monitor *Monitor, aggregator *aggregator.Aggregator, wg *sync.WaitGroup,
+	doneChan <-chan struct{}, ready *sync.Cond, logger *logrus.Entry,
+) {
+	// signal WaitGroup when goroutine finished
+	defer wg.Done()
+	defer monitor.Stop()
+
+	docsChan := make(chan proto.SystemProfile, collectorChanCapacity)
+	defer close(docsChan)
+
+	// monitor log file
+	monitor.Start(ctx, docsChan, doneChan, wg)
+
+	// signal we started monitoring
+	signalReady(ready)
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-doneChan:
+			return
+		default:
+		}
+
+		select {
+		case doc, ok := <-docsChan:
+			if !ok {
+				return
+			}
+
+			logger.Debugf("added to aggregator %v", doc.Query)
+			err := aggregator.Add(ctx, doc)
+			if err != nil {
+				logger.Warnf("couldn't add document to aggregator: %s", err)
+			}
+		case <-ctx.Done():
+			return
+		case <-doneChan:
+			return
+		}
```

Comment:
Will be refactor depends on result of discussion about:
https://perconadev.atlassian.net/browse/PMM-13966